File: Editorial/99_Recover_Binary_Search_Tree/C/3_Recursive_Inorder_Traversal.c
void findTwoSwapped(struct TreeNode *root, struct TreeNode **pred,
                    struct TreeNode **x, struct TreeNode **y) {
    if (root == NULL) return;
    findTwoSwapped(root->left, pred, x, y);
    if (*pred != NULL && root->val < (*pred)->val) {
        *y = root;
        if (*x == NULL)
            *x = *pred;
        else
            return;
    }
    *pred = root;
    findTwoSwapped(root->right, pred, x, y);
}
void recoverTree(struct TreeNode *root) {
    struct TreeNode *pred = NULL, *x = NULL, *y = NULL;
    findTwoSwapped(root, &pred, &x, &y);
    // Swap x and y values
    int tmp = x->val;
    x->val = y->val;
    y->val = tmp;
}

File: Editorial/0099_Recover_Binary_Search_Tree/CSharp/3_Recursive_Inorder_Traversal.cs
public class Solution {
    private TreeNode x = null, y = null, pred = null;

    private void FindTwoSwapped(TreeNode root) {
        if (root == null)
            return;
        FindTwoSwapped(root.left);
        if (pred != null && root.val < pred.val) {
            y = root;
            if (x == null)
                x = pred;
            else
                return;
        }

        pred = root;
        FindTwoSwapped(root.right);
    }

    public void RecoverTree(TreeNode root) {
        FindTwoSwapped(root);
        int tmp = x.val;
        x.val = y.val;
        y.val = tmp;
    }
}
